### PR TITLE
State for nPMI Plugin

### PR DESCRIPTION
**Pull request by Sparkier**
* Motivation for features / changes
Adding state functionality to nPMI plugin.

* Technical description of changes
Added store, reducers, selectors, and effects with the according tests.

* Screenshots of UI changes
No UI changes.

* Detailed steps to verify changes work correctly (as executed by you)
Can only run the tests, as the state is not used yet:
`bazel run //tensorboard/webapp:karma_test_chromium-local`

### CODE REVIEW

File: tensorboard/webapp/plugins/npmi/data_source/npmi_data_source.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  abstract fetchValues(): Observable<ValueListing>;
+}
+
+@Injectable()
+export class NpmiHttpServerDataSource implements NpmiDataSource {
+  private readonly httpPathPrefix = 'data/plugin/npmi';
+
+  constructor(private http: TBHttpClient) {}
+
+  fetchRuns() {
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/data_source/npmi_data_source.ts:39]: This was a left-over I forgot to remove. Removed it now.

File: tensorboard/webapp/plugins/npmi/effects/BUILD [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        "//tensorboard/webapp:app_state",
+        "//tensorboard/webapp/plugins/npmi/actions",
+        "//tensorboard/webapp/plugins/npmi/data_source",
+        "//tensorboard/webapp/plugins/npmi/store",
+        "//tensorboard/webapp/plugins/npmi/store:types",
+        "//tensorboard/webapp/plugins/npmi/testing",
+        "//tensorboard/webapp/webapp_data_source:http_client_testing",
+        "@npm//@angular/core",
+        "@npm//@ngrx/effects",
+        "@npm//@ngrx/store",
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/effects/BUILD:44]: Done.

File: tensorboard/webapp/plugins/npmi/effects/npmi_effects.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+} from './../store/npmi_selectors';
+import {
+  npmiLoaded,
+  annotationsRequested,
+  annotationsLoaded,
+  metricsRequested,
+  metricsLoaded,
+  valuesRequested,
+  valuesLoaded,
+} from './../actions';
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/effects/npmi_effects.ts:35]: Done.

File: tensorboard/webapp/plugins/npmi/effects/npmi_effects_test.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      lastLoadedTimeInMs: null,
+    });
+    store.overrideSelector(getValuesLoaded, {
+      state: DataLoadState.NOT_LOADED,
+      lastLoadedTimeInMs: null,
+    });
+    store.overrideSelector(getMetricsData, {
+      run_1: ['count@test', 'npmi@test'],
+    });
+    store.refreshState();
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/effects/npmi_effects_test.ts:85]: Done.

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+          ...state.valuesLoaded,
+          state: DataLoadState.FAILED,
+        },
+      };
+    }
+  ),
+  on(
+    actions.valuesLoaded,
+    (state: NpmiState, {values, metrics}): NpmiState => {
+      let countValuesData: ValueListing = {};
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts:166]: Done.

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        for (let row of values[key]) {
+          let countRow = [];
+          let npmiRow = [];
+          for (let index in metrics[key]) {
+            if (metrics[key][index].startsWith('count@')) {
+              countRow.push(row[index]);
+            } else if (metrics[key][index] === 'count') {
+              countData[key].push(row[index]);
+            } else if (metrics[key][index].startsWith('nPMI')) {
+              npmiRow.push(row[index]);
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts:182]: Done.

File: tensorboard/webapp/plugins/npmi/actions/result_actions.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+export const valuesRequested = createAction('[NPMI] nPMI Values Requested');
+
+export const valuesLoaded = createAction(
+  '[NPMI] nPMI Values Loaded',
+  props<{values: ValueListing; metrics: MetricListing}>()
+);
+
+export const valuesRequestFailed = createAction(
+  '[NPMI] nPMI Values Request Failed'
+);
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/actions/result_actions.ts:31]: Done.

File: tensorboard/webapp/plugins/npmi/data_source/npmi_data_source.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  abstract fetchAnnotations(): Observable<AnnotationListing>;
+  abstract fetchMetrics(): Observable<MetricListing>;
+  abstract fetchValues(): Observable<ValueListing>;
+}
+
+@Injectable()
+export class NpmiHttpServerDataSource implements NpmiDataSource {
+  private readonly httpPathPrefix = 'data/plugin/npmi';
+
+  constructor(private http: TBHttpClient) {}
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/data_source/npmi_data_source.ts:37]: Done.

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+      }
+      return {
+        ...state,
+        valuesData: values,
+        countValuesData: countValuesData,
+        npmiValuesData: npmiValuesData,
+        valuesLoaded: {
+          state: DataLoadState.LOADED,
+          lastLoadedTimeInMs: Date.now(),
+        },
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/store/npmi_reducers.ts:197]: For now, I removed the `valuesData` and `metricsData` as they are indeed not needed anymore. In the future I will do as we discussed and add data structures as needed for visualizations.

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    expect(nextState.annotationsLoaded.state).toBe(DataLoadState.LOADED);
+    expect(
+      nextState.annotationsLoaded.lastLoadedTimeInMs
+    ).toBeGreaterThanOrEqual(t0);
+    expect(nextState.annotationsData).toEqual({
+      run_1: ['annotation_1', 'annotation_2'],
+    });
+  });
+
+  it('Overrides existing annotations on successful annotations loading', () => {
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts:60]: lower case the `Overrides`

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import * as actions from '../actions';
+import {reducers} from './npmi_reducers';
+import {DataLoadState} from './npmi_types';
+import {createNpmiState} from '../testing';
+
+describe('Annotations loading', () => {
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts:20]: put everything describe in this file under `describe('npmi_reducers', () => { ... })`.

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+            'nPMI@test2',
+          ],
+        },
+        values: {
+          run_1: [
+            [3510517, 16719, 513767, 0.16871, -0.37206],
+            [1396813, 1896, 638967, 0.687616, 0.68116],
+          ],
+        },
+      })
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts:133]: Make this test interesting by varying the response. Few ideas:

1. `metrics['run_1'].length !== values['run_1'].length`
2. Change the order of the runs around. e.g., 
```
'count@test1',
'count',
'nPMI@test1',
'count@test2',
'nPMI@test2
```

File: tensorboard/webapp/plugins/npmi/effects/npmi_effects_test.ts [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+      fetchValuesSubject = new Subject();
+      fetchMetricsSpy = spyOn(dataSource, 'fetchMetrics').and.returnValue(
+        fetchMetricsSubject
+      );
+      fetchValuesSpy = spyOn(dataSource, 'fetchValues').and.returnValue(
+        fetchValuesSubject
+      );
+    });
+
+    it('loads Metrics and Values on plugin open if data is not loaded', () => {
```
Comment (stephanwlee) [inline tensorboard/webapp/plugins/npmi/effects/npmi_effects_test.ts:126]: test the case when one of two calls (metrics or values) fail.

File: tensorboard/webapp/plugins/npmi/effects/npmi_effects.ts [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+          map(() => void null)
+        );
+      })
+    );
+  }
+
+  private loadValues(prevStream$: Observable<unknown>) {
+    return prevStream$.pipe(
+      withLatestFrom(
+        this.store.select(getMetricsData),
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/effects/npmi_effects.ts]: Done.

[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/effects/npmi_effects.ts]: Done.

File: tensorboard/webapp/plugins/npmi/effects/npmi_effects_test.ts [hunk 15]
```diff
@@ -0,0 +1,147 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/effects/npmi_effects_test.ts:1]: Done.

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts (hunk 10 above)
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts:60]: Done

File: tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts (hunk 11 above)
Comment (Sparkier) [reply tensorboard/webapp/plugins/npmi/store/npmi_reducers_test.ts:20]: Done

**Pull request merged**